### PR TITLE
bus/nes_ctrl: Added support for Bandai Hyper Shot light gun.

### PR DESCRIPTION
**Pull request by 0kmg**
Software list items promoted to working (nes.xml)
---------------------------------------
Space Shadow (Japan)

### CODE REVIEW

**Comment by 0kmg** [comment]
Do I need to do anything about the overlapping default controls? Both the joypad direction inputs and the lightgun analog x/y inc/dec default to the keys i, j, k, l.

**Comment by 0kmg** [comment]
[Here](https://fcgamer.wordpress.com/2014/11/29/bandais-hyper-shot-gun-shootin-space-aliens-killin-time/) is a page with images of this controller and box. Basically it's a 3rd party gun with all the joypad buttons built in (sans the A button). It is required for exactly one game, but all the buttons and trigger work for any other Famicom games that read zapper or controller inputs from the expansion port AFAIK. Beyond what's in the PR I have no info about how else this may differ from the standard Famicom zapper.

File: src/devices/bus/nes_ctrl/zapper.cpp [hunk 1]
```diff
@@ -64,6 +99,12 @@ void nes_zapper_device::device_start()
 {
 }
 
+void nes_bandaihs_device::device_start()
+{
+	save_item(NAME(m_latch));
+	save_item(NAME(m_strobe));
+}
```
Comment (cuavas) [inline src/devices/bus/nes_ctrl/zapper.cpp:106]: You should really call `nes_zapper_device::device_start()` even if it does nothing.  It’s good practice to start the base device class.

File: src/devices/bus/nes_ctrl/zapper.cpp [hunk 2]
```diff
@@ -73,14 +114,20 @@ void nes_zapper_device::device_reset()
 {
 }
 
+void nes_bandaihs_device::device_reset()
+{
+	m_latch = 0xff;
+	m_strobe = 0xff;
+}
```
Comment (cuavas) [inline src/devices/bus/nes_ctrl/zapper.cpp:121]: Same for this one – call the base implementation even if it does nothing right now.

File: src/devices/bus/nes_ctrl/zapper.cpp (hunk 1 above)
Comment (0kmg) [reply src/devices/bus/nes_ctrl/zapper.cpp]: Done.

File: src/devices/bus/nes_ctrl/zapper.cpp (hunk 2 above)
Comment (0kmg) [reply src/devices/bus/nes_ctrl/zapper.cpp]: Done.

File: src/devices/bus/nes_ctrl/zapper.cpp [hunk 3]
```diff
@@ -73,14 +115,21 @@ void nes_zapper_device::device_reset()
 {
 }
 
+void nes_bandaihs_device::device_reset()
+{
+	nes_zapper_device::device_reset();
+	m_latch = 0xff;
+	m_strobe = 0xff;
```
Comment (cuavas) [inline src/devices/bus/nes_ctrl/zapper.cpp:122]: Does this device actually respond to a system reset, does the reset line not propagate to the controllers?  You shouldn’t really be resetting this stuff if the device doesn’t actually respond to a system reset directly.

Comment (0kmg) [reply src/devices/bus/nes_ctrl/zapper.cpp:122]: Well, this begs the question of why do any of the nes_ctrl devices respond to reset (they pretty much all do)? NES controller port pin outs at top left: https://wiki.nesdev.org/w/index.php?title=Controller_port_pinout. Famicom expansion port: https://wiki.nesdev.org/w/index.php?title=Expansion_port. Those don't look promising for peripherals responding to reset.

Comment (0kmg) [reply src/devices/bus/nes_ctrl/zapper.cpp]: Yes. I've fixed it now.

**Pull request merged**